### PR TITLE
[MRG][FIX] Fixes to examples

### PR DESCRIPTION
**Pull request by jaeilepp**
Noticed that with the new SPM datasets the examples are not working. See ``plot_spm_faces_datasets`` for example.
figure_1-6
figure_1-7
figure_1-8
figure_1-9
figure_1-10

### CODE REVIEW

**Comment by agramfort** [comment]
hum... what's wrong?

**Comment by jaeilepp** [comment]
Yeah. I can also look deeper later, but I was asking if anyone (@Eric89GXL ) has an idea...

**Comment by agramfort** [comment]
compare for now the entries in the meas info. We have utility function for
this.

**Comment by Eric89GXL** [comment]
I'm not sure what the difference would be. You could try converting the files using the C tools and swap in the .fif files for the CTF ones and see if you get the same or different results.

**Comment by jaeilepp** [comment]
Seems like with our python reader the data has some weird stuff at the start causing the events to be out of sync. This mess is not appearing with the converted fif file:
screenshot_2016-03-09_15-04-27

**Comment by Eric89GXL** [comment]
That looks like a bug. I wonder why it's not caught by the tests. Can you see if the verbose output is similar to what the C code says? CTF data has some skips, and they might not match properly. I probably won't have time to check for a while.

**Comment by jaeilepp** [comment]
Seems that the mess at the beginning is data from the end of the recording. I see similar patterns when comparing to the end of the c converted file. You can also see the reference signal on the first channel:
screenshot_2016-03-09_15-34-28

File: mne/io/ctf/ctf.py [hunk 1]
```diff
@@ -149,13 +150,13 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
         with open(self._filenames[fi], 'rb') as fid:
             for bi in range(len(r_lims)):
                 samp_offset = (bi + trial_start_idx) * si['res4_nsamp']
-                n_read = min(si['n_samp'] - samp_offset, si['block_size'])
+                n_read = min(si['n_samp_tot'] - samp_offset, si['block_size'])
```
Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: This was the bug. It wasn't caught by the tests because on our testing data ``n_samp`` and ``n_samp_tot`` are equal.

File: mne/io/ctf/ctf.py [hunk 2]
```diff
@@ -129,6 +129,7 @@ def __init__(self, directory, system_clock='truncate', preload=False,
             if sample_info['n_samp'] == 0:
                 break
             if len(fnames) == 0:
+                # XXX: Buffer size different to c version
                 info['buffer_size_sec'] = \
                     sample_info['block_size'] / info['sfreq']
```
Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: This is different from the c conversion tool. In effect it means that ``buffer_size_sec`` is different when reading with python version if ``n_trial`` == 1. I don't know the consequences of changing this so I didn't do it. Changing it also makes our tests fail.

File: mne/time_frequency/tfr.py [hunk 3]
```diff
@@ -505,7 +505,8 @@ def _preproc_tfr(data, times, freqs, tmin, tmax, fmin, fmax, mode,
     from ..viz.utils import _setup_vmin_vmax
 
     copy = baseline is not None
-    data = rescale(data, times, baseline, mode, copy=copy)
+    if mode is not None:
+        data = rescale(data, times, baseline, mode, copy=copy)
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: ``plot_stockwell`` failed earlier because of this.

File: examples/decoding/plot_decoding_time_generalization.py [hunk 4]
```diff
@@ -28,9 +28,9 @@
 # Preprocess data
 data_path = spm_face.data_path()
 # Load and filter data, set up epochs
-raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D_raw.fif'
+raw_path = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D.ds'
```
Comment (agramfort) [inline examples/decoding/plot_decoding_time_generalization.py]: I'd stick to raw_fname

File: examples/inverse/plot_covariance_whitening_dspm.py [hunk 5]
```diff
@@ -48,9 +48,9 @@
 data_path = spm_face.data_path()
 subjects_dir = data_path + '/subjects'
 
-raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D_raw.fif'
+raw_path = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D.ds'
 
-raw = io.Raw(raw_fname % 1, preload=True)  # Take first run
+raw = io.read_raw_ctf(raw_path % 1, preload=True)  # Take first run
```
Comment (agramfort) [inline examples/inverse/plot_covariance_whitening_dspm.py]: raw_fname

File: mne/io/ctf/ctf.py (hunk 1 above)
Comment (agramfort) [inline mne/io/ctf/ctf.py]: no need to update a test?

Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: As I said, our testing data has equal values for these. We should probably add dataset with different values. I wonder if @dengemann or someone else has nice small files.

Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: And I'm not entirely sure this is the correct solution to fix the bug. It almost looks like our python version tries to leave out some samples from the end (see https://github.com/mne-tools/mne-python/blob/master/mne/io/ctf/ctf.py#L214), but at least with our spm_faces dataset this wasn't behaving well. Maybe someone with deeper understanding of CTF format can help.

Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: And I'm not entirely sure this is the correct solution to fix the bug. It almost looks like our python version tries to leave out some samples from the end (see https://github.com/mne-tools/mne-python/blob/master/mne/io/ctf/ctf.py#L214), but at least with our spm_faces dataset this wasn't behaving well. Maybe someone with deeper understanding of CTF format can help.

**Comment by jaeilepp** [comment]
If we are content with the ctf fix, this is ready for review/merge. It can be fixed in another PR if needed.

File: mne/time_frequency/tfr.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: If mode is None it shouldn't do anything, why is this needed?

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Because it is not checked for ``None`` anywhere. You can try ``plot_stockwell`` on master and see why.

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Wait. The example is also a bit silly. It is setting the baseline and then saying that it shouldn't be used.

Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: It would be better to make the underlying function check then

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: To me, this seems like the appropriate place to do the checking. This can basically be rephrased to say ``if rescale: do rescale``. But the example is weird. What is the point of setting ``baseline=(-0.5, 0), mode=None``?

Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: There is already a check for "if baseline is None do nothing" so by leaving
this check here you're unnecessarily splitting the "do nothing" checks, and
potentially forcing similar code to go elsewhere in the future

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: So you're saying the baseline from the example should be removed?

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: And actually if this check is removed, it still won't work. The documentation says:
```
mode : None | 'logratio' | 'ratio' | 'zscore' | 'mean' | 'percent'
            Do baseline correction with ratio (power is divided by mean
            power during baseline) or zscore (power is divided by standard
            deviation of power during baseline after subtracting the mean,
            power = [power - mean(power_baseline)] / std(power_baseline)).
            If None no baseline correction is applied.
```

Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: As far as the code here is concerned, I'm saying that the check that is currently here should be moved to `rescale` since rescale already has related checks.

As far as the code in the example is concerned, we should do whatever makes the example look most reasonable. You could try setting `mode` to different values and see what works best.

Come to think of it, we should probably make it throw an error if `baseline is not None and mode is None` to avoid people thinking they're baseline correcting when they're not. And those checks should be done in `rescale`.

File: mne/io/ctf/ctf.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: This sets what the block size will be when writing to disk, so it shouldn't make a big difference either way. There was a reason I did it this way, but I don't remember what it was offhand. Maybe we actually use it while reading (which wouldn't be great)? Anyway I wouldn't bother changing it.

File: mne/io/ctf/ctf.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: Both the Python and C versions can omit samples from the end. We do it differently from the C tools to handle a user's use case, see:

https://github.com/mne-tools/mne-python/pull/2712

I suspect this fix will break that use case, but I'm not sure. We really do need to add some test files to figure it out.

File: mne/io/ctf/ctf.py (hunk 2 above)
Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: Okay, seems that the bug is maybe different to what I thought at first. With ``system_clock='ignore'`` the data seems to load alright, but the stim channel is out of sync. I'll take a closer look tomorrow.

File: mne/time_frequency/tfr.py (hunk 3 above)
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: > Come to think of it, we should probably make it throw an error if baseline is not None and mode is None to avoid people thinking they're baseline correcting when they're not. And those checks should be done in rescale.

Agreed.

File: mne/io/ctf/ctf.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: Thanks for looking into it. It will be good for you to take a look instead of me so that more than one person has a handle on what the code is doing.

Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: Sorry, I was looking at the wrong stim channel. Maybe there isn't a bug. Will check tomorrow...

**Comment by agramfort** [comment]
I would check against brainstorm reader to make sure we read the correct number of samples and proper values of everything...

**Comment by Eric89GXL** [comment]
You'll need a rebase since I decreased mem usage in the SPM example

**Comment by jaeilepp** [comment]
> Come to think of it, we should probably make it throw an error if baseline is not None and mode is None to avoid people thinking they're baseline correcting when they're not. And those checks should be done in rescale.

Now this happens. It might break someone's code, but that means they've been doing it wrong all along.

To me it looks like the ctf bug fix here is the right one. Calling with ``system_clock='ignore'`` on master makes the artefacts in the beginning disappear, but it also added samples to the end, which doesn't happen with c version. With this fix, 'ignore' adds those samples too, but the artefacts aren't there either way. So judging by the behavior, the fix works.

**Comment by Eric89GXL** [comment]
Yeah the C code doesn't have an option to ignore to clock IIRC. So now the
raw data looks correct, the example renders properly, and tests pass?

Can you add a test that requires the spm face dataset to catch the
regression? It won't run on Travis but it will on most dev machines.

**Comment by jaeilepp** [comment]
Okay test added. Now it just checks that samples are omitted. For data comparison, the data should be added to the dataset. I believe this is sufficient for now.
 I also removed special characters from maxwell.py. They were causing failures with sphinx-gallery.
Ready for review from my end.

**Comment by Eric89GXL** [comment]
Can we improve sphinx gallery instead of removing those characters? There
are multiple other files that have them as well, and they are convenient to
have. Probably better to open an issue to see if we can deal with it
properly instead of tacking it on here.

File: examples/visualization/plot_clickable_image.py [hunk 6]
```diff
@@ -24,10 +25,12 @@
...
 # Set parameters and paths
 plt.rcParams['image.cmap'] = 'gray'
-im_path = op.join(op.dirname(mne.__file__), 'data', 'image', 'mni_brain.gif')
+
+im_path = op.join(op.dirname(__file__), '..', '..', 'mne', 'data', 'image',
+                  'mni_brain.gif')
 
 # We've already clicked and exported
-layout_path = op.join(op.dirname(mne.__file__), 'data', 'image')
+layout_path = op.join(op.dirname(__file__), '..', '..', 'mne', 'data', 'image')
```
Comment (agramfort) [inline examples/visualization/plot_clickable_image.py]: you cannot assume the example file is at a fixed place on disk. The previous option is more robust.

File: examples/time_frequency/plot_stockwell.py [hunk 7]
```diff
@@ -45,6 +45,6 @@
 power, itc = tfr_stockwell(epochs, fmin=6., fmax=30., decim=4, n_jobs=1,
                            width=.3, return_itc=True)
 
-power.plot([0], baseline=(-0.5, 0), mode=None, title='S-transform (power)')
+power.plot([0], baseline=None, mode=None, title='S-transform (power)')
```
Comment (agramfort) [inline examples/time_frequency/plot_stockwell.py]: why?

File: mne/io/ctf/tests/test_ctf.py [hunk 8]
```diff
@@ -197,4 +197,15 @@ def test_read_ctf():
     assert_raises(ValueError, read_raw_ctf,
                   op.join(ctf_dir, ctf_fname_continuous), 'foo')
 
+
+@testing.requires_spm_data
+def test_read_spm_ctf():
+    """Test CTF reader with omitted samples."""
+    data_path = spm_face.data_path()
+    raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces1_3D.ds'
```
Comment (agramfort) [inline mne/io/ctf/tests/test_ctf.py]: use os.path.join in tests.

Comment (agramfort) [inline mne/io/ctf/tests/test_ctf.py]: for our windows users...

**Comment by agramfort** [comment]
I agree with @Eric89GXL.

the bug should be fixed in sphinx-gallery.

Please open an issue there and cc me.

thx

File: examples/time_frequency/plot_stockwell.py (hunk 7 above)
Comment (Eric89GXL) [inline examples/time_frequency/plot_stockwell.py]: From the long discussion before, it wasn't doing anything before, this
makes it more explicit (mode=None does nothing) and avoids what is now an
error (mode=None with a baseline not None)

**Comment by jaeilepp** [comment]
To me allowing special characters sounds like asking for trouble. The doc builds are way too fragile to some specific versions as it is.

**Comment by agramfort** [comment]
at least raise the issue on sphinx gallery. They are really responsive.

**Comment by Eric89GXL** [comment]
UTF8 is pretty standard (and default on Py3k) it should really be
supported. I'd rather not mangle our docstrings if we don't have to

File: mne/baseline.py [hunk 9]
```diff
@@ -55,15 +59,12 @@ def rescale(data, times, baseline, mode='mean', copy=True, verbose=None):
     data_scaled: array
         Array of same shape as data after rescaling.
     """
+    _log_rescale(baseline, mode)
+    if baseline is None:
+        return data
```
Comment (Eric89GXL) [inline mne/baseline.py]: This changes the previous reasonable behavior of the function -- before if `copy is True` then it would always return a copy, now it does in some cases and doesn't in others.

File: mne/baseline.py [hunk 10]
```diff
@@ -55,15 +59,12 @@ def rescale(data, times, baseline, mode='mean', copy=True, verbose=None):
...
     if copy:
         data = data.copy()
 
-    valid_modes = ('logratio', 'ratio', 'zscore', 'mean', 'percent',
-                   'zlogratio')
-    if mode not in valid_modes:
-        raise Exception('mode should be any of : %s' % (valid_modes, ))
-
-    _log_rescale(baseline, mode)
     if baseline is not None:
```
Comment (Eric89GXL) [inline mne/baseline.py]: you no longer need this conditional because `if baseline is None` it has returned already

File: mne/io/ctf/tests/test_ctf.py [hunk 11]
```diff
@@ -197,4 +197,15 @@ def test_read_ctf():
     assert_raises(ValueError, read_raw_ctf,
                   op.join(ctf_dir, ctf_fname_continuous), 'foo')
 
+
+@testing.requires_spm_data
```
Comment (Eric89GXL) [inline mne/io/ctf/tests/test_ctf.py]: this should probably be `spm_face.requires_spm_data`. `testing` is the namespace for the `testing` dataset, not all things related to testing

File: mne/io/ctf/tests/test_ctf.py [hunk 12]
```diff
@@ -197,4 +197,15 @@ def test_read_ctf():
...
+def test_read_spm_ctf():
+    """Test CTF reader with omitted samples."""
+    data_path = spm_face.data_path()
+    raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces1_3D.ds'
+    raw = read_raw_ctf(raw_fname)
+    extras = raw._raw_extras[0]
+    assert_equal(extras['n_samp'], raw.n_times)
+    assert_false(extras['n_samp'] == extras['n_samp_tot'])
```
Comment (Eric89GXL) [inline mne/io/ctf/tests/test_ctf.py]: Probably also check that `len(raw.times) == extras['n_samp_tot']` or whatever the correct conditional is? Does this test fail on `master` (it should)
?

**Comment by Eric89GXL** [comment]
How is it breaking rendering for you? The (TM) seems to work fine for me (second warning box):

screenshot from 2016-03-11 11-56-20

File: mne/io/ctf/tests/test_ctf.py (hunk 12 above)
Comment (jaeilepp) [inline mne/io/ctf/tests/test_ctf.py]: Actually that is tested on the line above ``assert_equal(extras['n_samp'], raw.n_times)``.

**Comment by jaeilepp** [comment]
> How is it breaking rendering for you?

``plot_maxwell`` fails on build stage on dev version of sphinx-gallery.

**Comment by Eric89GXL** [comment]
Is there a way to build individual examples now? I thought you and
@jasmainak worked something out. My mayavi is broken so I can't build them
all :(

**Comment by Eric89GXL** [comment]
In any case, I think this is good to merge if you get rid of the maxwell
changes

**Comment by jasmainak** [comment]
yes, you can build individual examples. Just do:

```py
make html_dev-pattern PATTERN='/decoding/plot_decoding.py'
```
PATTERN can be a regexp

**Comment by Eric89GXL** [comment]
Awesome, thanks

**Comment by agramfort** [comment]
let me know if we're good here.

**Comment by Eric89GXL** [comment]
This should fix `sphinx-gallery`:

https://github.com/sphinx-gallery/sphinx-gallery/pull/106

File: examples/visualization/plot_clickable_image.py (hunk 6 above)
Comment (jaeilepp) [inline examples/visualization/plot_clickable_image.py]: It fails on circle. That's why I changed it.

**Comment by jaeilepp** [comment]
Ready for review.

File: examples/inverse/plot_covariance_whitening_dspm.py [hunk 13]
```diff
@@ -48,12 +48,12 @@
 data_path = spm_face.data_path()
 subjects_dir = data_path + '/subjects'
 
-raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D_raw.fif'
+raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D.ds'
 
-raw = io.Raw(raw_fname % 1)  # Take first run
+raw = io.read_raw_ctf(raw_fname % 1, preload=True)  # Take first run
```
Comment (Eric89GXL) [inline examples/inverse/plot_covariance_whitening_dspm.py]: You shouldn't need `preload=True`, if you do, we have a bug

Comment (Eric89GXL) [inline examples/inverse/plot_covariance_whitening_dspm.py]: (this will increase memory usage substantially)

File: examples/visualization/plot_clickable_image.py (hunk 6 above)
Comment (Eric89GXL) [inline examples/visualization/plot_clickable_image.py]: What's the failure? Is there some way to use the location of `mne` instead of the location of the example file? It really should be more robust to use the repo-location to find data files from the repo than use the example-location -- people could reasonably just download the example to their Desktop and expect it to work. Maybe we need to add `mne.data_path` within the repo that just points to `op.join(os.abspath(__file__), 'data')` or so in `__init__.py`.

File: mne/datasets/testing/_testing.py [hunk 14]
```diff
@@ -12,6 +12,7 @@
 
 
 has_testing_data = partial(has_dataset, name='testing')
+has_spm_data = partial(has_dataset, name='spm')
```
Comment (Eric89GXL) [inline mne/datasets/testing/_testing.py]: shouldn't be in `_testing`

Comment (Eric89GXL) [inline mne/datasets/testing/_testing.py]: (`_testing` namespace here is for the testing dataset, not all testing-related-things)

File: examples/visualization/plot_clickable_image.py (hunk 6 above)
Comment (jaeilepp) [inline examples/visualization/plot_clickable_image.py]: See https://circle-artifacts.com/gh/mne-tools/mne-python/451/artifacts/1/home/ubuntu/mne-python/doc/_build/html/auto_examples/visualization/plot_clickable_image.html#sphx-glr-auto-examples-visualization-plot-clickable-image-py

File: examples/inverse/plot_covariance_whitening_dspm.py (hunk 13 above)
Comment (jaeilepp) [inline examples/inverse/plot_covariance_whitening_dspm.py]: Whoops. I don't know how that got there. But in any case there is a ``load_data`` on the line below so this seems redundant.

Comment (Eric89GXL) [inline examples/inverse/plot_covariance_whitening_dspm.py]: More importantly, the `load_data` happens *after* the `crop` statement, which leads to memory savings

File: examples/visualization/plot_clickable_image.py (hunk 6 above)
Comment (agramfort) [inline examples/visualization/plot_clickable_image.py]: I suspect that the gif file is not properly installed via our setup.py .... it's a bug. The example should not be changed due to this.

Comment (Eric89GXL) [inline examples/visualization/plot_clickable_image.py]: Good call, it is indeed missing

https://github.com/mne-tools/mne-python/blob/master/setup.py#L106

File: mne/io/ctf/tests/test_ctf.py [hunk 15]
```diff
@@ -197,4 +197,16 @@ def test_read_ctf():
     assert_raises(ValueError, read_raw_ctf,
                   op.join(ctf_dir, ctf_fname_continuous), 'foo')
 
+
+@spm_face.requires_spm_data
+def test_read_spm_ctf():
+    """Test CTF reader with omitted samples."""
+    data_path = spm_face.data_path()
+    raw_fname = data_path + op.join('MEG', 'spm',
```
Comment (Eric89GXL) [inline mne/io/ctf/tests/test_ctf.py]: why not just `op.join(data_path, 'MEG', ...)`?

Comment (agramfort) [inline mne/io/ctf/tests/test_ctf.py]: indeed

**Comment by agramfort** [comment]
besides LGTM

don't forget to update what's new BUG section

**Comment by jaeilepp** [comment]
I thought I had found a setting that would work with mayavi figures, but apparently I was wrong. I'll try it on another PR then.
Ready for review.

**Comment by agramfort** [comment]
Great

Merge when Travis is happy
